### DISCUSSION TITLE
How to get working in static folders

### DISCUSSION THREAD

**Discussion by Alerinos** [SixLabors/ImageSharp.Web#144] [Q&A] [answered]
Hi, I have some additional static folders set up
```
            app.UseImageSharp();    // Add the image processing middleware.

            app.UseHttpsRedirection();  // wwwroot
            app.UseStaticFiles(new StaticFileOptions {
                OnPrepareResponse = ctx => {
                    ctx.Context.Response.Headers.Append(
                        "Cache-Control", $"public, max-age={cacheMaxAge}");
                }
            });

            // Public folder
            app.UseStaticFiles(new StaticFileOptions {
                OnPrepareResponse = ctx => {
                    ctx.Context.Response.Headers.Append(
                        "Cache-Control", $"public, max-age={cacheMaxAge}");
                },
                FileProvider = new PhysicalFileProvider(
                Path.Combine(Directory.GetCurrentDirectory(), "Public")),
                RequestPath = "/Public",
            });
```
The wwwroot folder works fine, but how to support the "Public" folder?

error console:
```
SixLabors.ImageSharp.Web.Middleware.ImageSharpMiddleware[2]
      The image 'https://localhost:5001/Public/news/67a8376d-6bcc-4d32-b191-7c728a02df67/source.png?width=200&height=50&rsampler=nearest&rmode=stretch' could not be resolved
```

**Comment by JimBobSquarePants** [comment]
The current provider is tied to the wwwroot folder via the `WebRootFileProvider` so cannot support multiple static folders.
https://github.com/SixLabors/ImageSharp.Web/blob/7ab73bb28f7eee3c16244ef343af9a5bf2291ea6/src/ImageSharp.Web/Providers/PhysicalFileSystemProvider.cs#L43

The best solution would be to implement your own provider that took an collection of relative root folder names as options and created a `PhysicalFileProvider` for each one to check against. 

We actually do something similar for the physical cache (albeit with only a single folder)

https://github.com/SixLabors/ImageSharp.Web/blob/7ab73bb28f7eee3c16244ef343af9a5bf2291ea6/src/ImageSharp.Web/Caching/PhysicalFileSystemCache.cs#L81

If you are willing to PR against the library I can help you with the development which we can then use to replace the existing functionality in a backwards compatible manner.

**Comment by deanmarcussen** [comment, marked as answer]
You should be able to use a `CompositeFileProvider` to do this.

In your startup something like this.
```
env.WebRootFileProvider = new CompositeFileProvider(env.WebRootFileProvider, new PhysicalFileProvider("/Public"));
```

Then you don't need two `UseStaticFiles()` and ImageSharp will just pickup on the changed configuration, when it reads from

**Reply by sebastienros** [reply]
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/file-providers?view=aspnetcore-5.0#compositefileprovider

**Reply by JimBobSquarePants** [reply]
Nice! I didn't know you could do that. Thanks guys, this is way easier!

**Reply by Alerinos** [reply]
It works, thanks! Now you have to delete "Public" from the link everywhere. Possibly, it is possible to add a "dynamic" name in the link so that the url does not need to be changed?

**Reply by deanmarcussen** [reply]
That's just a matter of setting a folder root on the `PhysicalFileProvider`

https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.fileproviders.physicalfileprovider.-ctor?view=dotnet-plat-ext-5.0#Microsoft_Extensions_FileProviders_PhysicalFileProvider__ctor_System_String_

**Reply by xps** [reply]
Hi @deanmarcussen, can you please explain what you meant by "setting a folder root on the PhysicalFileProvider"?
I'm also trying to map a "Public" folder (which is outside of wwwroot) to the "/public" URL, and still have ImageSharp process the images in that folder. I did not manage to use the CompositeFileProvider while keeping the URL structure.
